src/openforms/submissions/tasks/__init__.py

```diff
@@ -0,0 +1,82 @@
+    register_submission_task = register_submission.si(submission_id)
+    obtain_submission_reference_task = obtain_submission_reference.si(submission_id)
+    finalize_completion_task = finalize_completion.si(submission_id)
+
+    # for the orchestration with distributed processing and dependencies between
+    # tasks, see the Celery documentation:
+    # https://docs.celeryproject.org/en/stable/userguide/canvas.html#guide-canvas
+    # Additionally, check the developer documentation of Open Forms for a visual
+    # representation of the task orchestration and dependencies (TODO, @joeri)
+    # The linked task (= next task) is only executed if the previous task returns
+    # successfully, so error handling needs to happen inside each task.
+    on_completion_chain = chain(
+        # The appointment must be registered before a confirmation PDF is generated and
+        # any backend registration happens, as on-failure, the user should get feedback
+        # about the failure.
+        register_appointment_task,
+        # The submission report needs to already have been generated before it can be
+        # attached in the registration backend.
+        # TODO: can be in parallel with register_appointment_task and if that fails -> delete the report again
+        generate_report_task,
+        # TODO: ensure that any images that need resizing are done so before this is attempted
+        register_submission_task,
+        obtain_submission_reference_task,
+        update_appointment_task,
+        # we schedule the finalization so that the ``async_result`` below is marked
+        # as done, which is the "signal" to show the confirmation page. Actual payment
+        # flow & confirmation e-mail follow later.
+        finalize_completion_task,
+    )
+
+    # this can run any time because they have been claimed earlier
+    cleanup_temporary_files_for.delay(submission_id)
+
+    async_result: AsyncResult = on_completion_chain.delay()
+    # NOTE - this is "risky" since we're running outside of the transaction (this code
+    # should run in transaction.on_commit)!
+    Submission.objects.filter(id=submission_id).update(
+        on_completion_task_id=async_result.id
+    )
+    return async_result.id
+
+
+@app.task(bind=True)
+def finalize_completion(task, submission_id: int) -> None:
+    """
+    Schedule all the tasks that need to happen to finalize the submission completion.
+
+    Finalization happens _after_ the confirmation screen is shown to the end-user.
+    Showing this screen depends on all the previous registration tasks being completed,
+    so the :func:`on_completion` handler must kick this off AND have its own task ID
+    that finishes, which is checked in the submission status endpoint.
+    """
+    send_confirmation_email_task = maybe_send_confirmation_email.si(submission_id)
+    send_confirmation_email_task.delay()
+
+    # TODO: clean up session data
```

Comment:
likely not here and rather keep it in the request-response cycle, but leaving the TODO here for now as a reminder.

Reply:
Why don't we just cleanup the session when the confirmation page is shown (aka status=done, result=success)?

This also solves the issue for adding the submission to the session again, mentioned in a TODO earlier.

If we need to block double submissions, i rather fix it with a check on the submission status.

Reply:
because that requires the API client to have called the status endpoint or call an endpoint-to-be-added, and we can't assume that all API clients will behave the way they should. If we keep this server-side, it's guaranteed to work the way it should.

Especially with the timestamped HMAC token approach we have fairly simple options to manage that.

Removing the submission from the session is important to make the submission data immutable, as otherwise it would be possible to mutate state and have the submission data be different from what is sent to the registration backend. I just want to block that completely - you should not have access to the step data endpoints at all anymore.